pkg/appgw/frontend_listeners.go

```diff
@@ -1,16 +1,93 @@
+// getFrontendListeners constructs the unique set of App Gateway HTTP listeners across all ingresses.
+func (builder *appGwConfigBuilder) getFrontendListeners(ingressList []*v1beta1.Ingress) (*[]n.ApplicationGatewayHTTPListener, map[frontendListenerIdentifier]*n.ApplicationGatewayHTTPListener) {
+	// TODO(draychev): this is for compatibility w/ RequestRoutingRules and should be removed ASAP
+	legacyMap := make(map[frontendListenerIdentifier]*n.ApplicationGatewayHTTPListener)
+
+	var httpListeners []n.ApplicationGatewayHTTPListener
+
+	for listener, config := range builder.getListenerConfigs(ingressList) {
+		var secretFullName string
+		var protocol n.ApplicationGatewayProtocol
+
+		if config != nil {
+			protocol = config.Protocol
+			secretFullName = config.Secret.secretFullName()
+		} else {
+			// Default protocol
+			protocol = n.HTTP
+		}
+
+		httpListener := builder.newAppGatewayHTTPListener(listener, protocol)
+
+		listenerHasHostname := len(*httpListener.ApplicationGatewayHTTPListenerPropertiesFormat.HostName) > 0
+
+		if protocol == n.HTTPS {
+			sslCertificateName := secretFullName
+			sslCertificateID := builder.appGwIdentifier.sslCertificateID(sslCertificateName)
+			httpListener.SslCertificate = resourceRef(sslCertificateID)
+
+			if listenerHasHostname {
+				httpListener.RequireServerNameIndication = to.BoolPtr(true)
+			}
+		}
+
+		if listenerHasHostname {
+			// Put the listener at the front of the list!
+			httpListeners = append([]n.ApplicationGatewayHTTPListener{httpListener}, httpListeners...)
+		} else {
+			httpListeners = append(httpListeners, httpListener)
+		}
 
- 	for _, ingress := range ingressList {
-		_, _, configMap := builder.processIngressRules(ingress)
-		for k, v := range configMap {
-			httpListenersAzureConfigMap[k] = v
+		legacyMap[listener] = &httpListener
+	}
+
+	// TODO(draychev): The second map we return is for compatibility w/ RequestRoutingRules and should be removed ASAP
+	return &httpListeners, legacyMap
+}
+
+// getListenerConfigs creates an intermediary representation of the listener configs based on the passed list of ingresses
+func (builder *appGwConfigBuilder) getListenerConfigs(ingressList []*v1beta1.Ingress) map[frontendListenerIdentifier]*frontendListenerAzureConfig {
+	allListeners := make(map[frontendListenerIdentifier]*frontendListenerAzureConfig)
+	for _, ingress := range ingressList {
+		_, azListenerConfigs := builder.processIngressRules(ingress)
+		for listenerID, azConfig := range azListenerConfigs {
+			allListeners[listenerID] = azConfig
 		}
 	}
 
- 	return httpListenersAzureConfigMap
+	// App Gateway must have at least one listener - the default one!
+	if len(allListeners) == 0 {
+		allListeners[defaultFrontendListenerIdentifier()] = nil
+	}
+
+	return allListeners
+}
+
+func (builder *appGwConfigBuilder) newAppGatewayHTTPListener(listener frontendListenerIdentifier, protocol n.ApplicationGatewayProtocol) n.ApplicationGatewayHTTPListener {
```

Comment:
Can we rename this to `newFrontEndListener`. Will be consistent with the file name. Also `AppGateway` here seems a bit redundant. or `newHTTPListener` would work as well. A bit inconsistent, but I think that comes from our naming convention in Application Gateway as well. Or you could change all `FrontEndListener` to `HTTPListener`? A bit confusing with these terms.

Reply:
It is indeed a struggle to keep the naming convention consistent. 
After this refactor is done - I'll take a broader look at whether we really need all the `internaltypes` we have that reference a `listener` in some way.

So far we have:
  - `frontendListenerIdentifier` - https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/pkg/appgw/internaltypes.go#L30-L33

  - `frontendListenerAzureConfig ` - https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/pkg/appgw/internaltypes.go#L50-L54  (this seems redundant)

  - and then of course the SDK's `ApplicationGatewayHTTPListener`